src/psyclone/psyir/nodes/psy_data_node.py

```diff
@@ -0,0 +1,536 @@
+
+        # Store the name of the PSyData variable that is used for this
+        # PSyData name. This allows to show the variable name in __str__
+        # (and also if we would call create_name in gen(), the name would
+        # change every time gen() is called).
+        from psyclone.psyGen import NameSpaceFactory
+        self._var_name = NameSpaceFactory().create().create_name("psy_data")
+
+        if not children:
+            super(PSyDataNode, self).__init__(ast=ast, children=children,
+                                              parent=parent)
+        else:
+            node_parent = children[0].parent
+            node_position = children[0].position
+
+            # A PSyData node always contains a Schedule
+            sched = self._insert_schedule(children)
+
+            super(PSyDataNode, self).__init__(ast=ast, children=[sched],
+                                              parent=parent)
+
+            # Correct the parent's list of children. Use a slice of the list
+            # of nodes so that we're looping over a local copy of the list.
+            # Otherwise things get confused when we remove children from
+            # the list.
+            for child in children[:]:
+                # Remove child from the parent's list of children
+                node_parent.children.remove(child)
+
+            # Add this node as a child of the parent
+            # of the nodes being enclosed and at the original location
+            # of the first of these nodes
+            node_parent.addchild(self, index=node_position)
+
+        # Name and colour to use for this node - must be set after calling
+        # the constructor
+        self._text_name = "PSyData"
+        self._colour_key = "PSyData"
+
+        # Name of the region. In general at constructor time we might
+        # not have a parent subroutine or a child for the kernel, so
+        # the name is left empty, unless explicitly provided by the
+        # user. If names are not provided here then the region and
+        # module names are set the first time gen() is called (and
+        # then remain unchanged).
+        self._module_name = None
+        self._region_name = None
+        # The region identifier caches the computed module- and
+        # region-name as a tuple of strings
+        self._region_identifier = ("", "")
+
+        if name:
+            # pylint: disable=too-many-boolean-expressions
+            if not isinstance(name, tuple) or not len(name) == 2 or \
+               not name[0] or not isinstance(name[0], str) or \
+               not name[1] or not isinstance(name[1], str):
+                raise InternalError(
+                    "Error in PSyDataNode. The name must be a "
+                    "tuple containing two non-empty strings.")
+            # pylint: enable=too-many-boolean-expressions
+            # Valid PSyData names have been provided by the user.
+            self._module_name = name[0]
```

Comment:
Hmm - we originally added the two names (module and region name or so) based on the idea that some profiling libraries (maybe drhook??? not sure now) would support that (and allow grouping).

Now imho when I want to set a name as a normal user, I just want to provide one string ("stupid-region-that-Im-looking-at"), but Rupert sensibly supported two strings. Since there is no API to set/get the module/region names (it's only used in derived classes, directly accessing self._module), but I preferred a function call ... I added this function, but .. am happy to use a different name.

Reply:
Looking at this again, I think what I meant was:

Why bother with a `set_region_identifier`? We can just have `region_identifier` do:

    return (self._module_name, self._region_name)

and not bother with `self._region_identifier`.

Reply:
This one still needs to be addressed...

Reply:
Ha ... that one ... sorry, I missed that when I went back trying to answer all questions - my idea that not answering the 'simple typo' comments to reduce emails backfired here, I missed that.

I originally implemented this obviously the way you said, but I ran into a failing test case: test_region in profile test. This test adds a first profiling region, calls ``gen()`` , adds a secpmd profiling region and calls ``gen()`` again. Now if we store the computed module (and region) name in self._module_name in the first gen() call, the second time ``gen()`` is called, it will detect that self._module_name is not empty, and use this name (since it looks like a user defined name now). No big problem so far, but:
the way @rupertford implemented the algorithm to determine a unique name means that the second time ``gen()`` is called (i.e. when there are two profiling regions), the first profiling region gets a different name. Even worse, the then computed name for the second profiling region is the same as the stored name for the first profiling region - so we have a duplicated module (and region) name.

I contacted Rupert about this, and he told me that the name should reflect the structure of the tree, and if the tree changes, the names can change. And so we can't store a computed name in self._module_name.

I've added a long description of this issue as comment. Sorry for missing this.

Reply:
Ah, finally light dawns for me :-) Thanks for explaining and adding the comment.